operator: Migrate Cilium Endpoint GC to hive

alan-kut opened the pull request:
operator: Migrate Cilium Endpoint GC to hive

```release-note
```

alan-kut committed:
operator/resources: Add CiliumNode resource

Signed-off-by: Alan Kutniewski <kutniewski@google.com>

alan-kut committed:
operator/resources: Add Pod resource

Signed-off-by: Alan Kutniewski <kutniewski@google.com>

alan-kut committed:
operator/endpointgc: Migrate CEP GC to hive

Signed-off-by: Alan Kutniewski <kutniewski@google.com>
Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

pippolo84 commented on operator/endpointgc/cell.go:
Nit: this file and `gc.go` have been added empty in the second commit. Can you please rebase it so they will be added only in third commit?

alan-kut replied:
Done

pippolo84 commented on operator/endpointgc/cell.go:
```diff
@@ -1 +1,43 @@
+	// attempted to be collected
+	Interval = "cilium-endpoint-gc-interval"
+)
+
```
This is unused, so we can remove it.

alan-kut replied:
Removed, thanks.
Firstly I wanted to move the flag to the gc but it is also used a little bit in the root. I decided to use shard config but forgot to remove everything

pippolo84 commented on operator/endpointgc/gc.go:
```diff
@@ -1 +1,266 @@
+		})
+	switch {
+	case err == nil:
+		metrics.EndpointGCObjects.WithLabelValues(metrics.LabelValueOutcomeSuccess).Inc()
```
We recently introduced support for modular metrics in the operator https://github.com/cilium/cilium/pull/28005. We can now use `cell.Metric` to declare cell-specific metrics instead of using the legacy (global) ones. Do you mind modularize the metrics too?
You can find an example of how to do so [here](https://github.com/cilium/cilium/pull/28166/commits/62ef3f46d776b1203f7be1b0ab12d146d55e3f62) (CES controller metrics modularization)

alan-kut replied:
Done

pippolo84 reviewed (commented):
This is nice, thanks! :100:
I've left a couple of nits inline and a suggestion about metrics, but overall LGTM.
I didn't review the tests yet, I aim to do that in a second pass.

alan-kut committed:
operator/endpointgc: Use modular metrics from hive pkg

Use modular metrics instead of legacy ones for the endpoint GC.
Update also the operator API server metrics test accordingly.

Signed-off-by: Alan Kutniewski <kutniewski@google.com>
Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

pippolo84 commented on operator/api/metrics_test.go:
You also need to remove the check on `cilium_operator_endpoint_gc_objects` in `TestMetricsHandlerWithMetrics`:

```diff
-       if err := testMetric(
-               metrics,
-               "cilium_operator_endpoint_gc_objects",
-               float64(1),
-               map[string]string{
-                       operatorMetrics.LabelOutcome: operatorMetrics.LabelValueOutcomeSuccess,
-               },
-       ); err != nil {
-               t.Fatalf("error while inspecting metric: %s", err)
-       }
```

otherwise the test fails:

```
--- FAIL: TestMetricsHandlerWithMetrics (0.45s)
    metrics_test.go:189: error while inspecting metric: "cilium_operator_endpoint_gc_objects" not found
```

alan-kut replied:
Fixed

pippolo84 commented on operator/endpointgc/gc_test.go:
```diff
@@ -0,0 +1,246 @@
+func waitForCEPs(t *testing.T, cepStore resource.Store[*cilium_v2.CiliumEndpoint], number int) {
+	if err := testutils.WaitUntil(func() bool {
+		return len(cepStore.List()) == number
+	}, waitTime*5); err != nil {
```
I suggest to increase this a bit, maybe to a total of 5 or 10 seconds. This is to reduce flakyness in CI, where the performance characteristic is so unpredictable. Also, this won't introduce any fixed delay, since the condition is still checked every 10ms.

alan-kut replied:
Changed. I have a problem with a test for disabled endpoints gc.
```go
// wait for all CEPs created by prepareCiliumEndpoints
waitForCEPs(t, cepStore, 8)
// wait for potential GC
time.Sleep(500 * time.Millisecond)
// gc is disabled so no CEPs should be deleted
waitForCEPs(t, cepStore, 8)
```
It's hard to tell how long should we wait to make sure gc is really disabled.

But maybe it's fine to just remove this test.

pippolo84 replied:
> It's hard to tell how long should we wait to make sure gc is really disabled.

That's right.
In this case, I don't think we can do much to improve this. Maybe what we can do is explicitly setting the gc interval with a value much smaller than the observation interval:

```diff
diff --git a/operator/endpointgc/gc_test.go b/operator/endpointgc/gc_test.go
index d0e23d52e5..d3acc2a8ee 100644
--- a/operator/endpointgc/gc_test.go
+++ b/operator/endpointgc/gc_test.go
@@ -112,6 +112,7 @@ func TestRegisterControllerWithCRDDisabled(t *testing.T) {
                cell.Metric(NewMetrics),
                cell.Provide(func() SharedConfig {
                        return SharedConfig{
+                               Interval:                 100 * time.Millisecond,
                                DisableCiliumEndpointCRD: true,
                        }
                }),
```

Unfortunately, this is still time-dependent, but (hopefully) should be fine in most cases and makes the test intent a bit clearer.

pippolo84 commented on operator/endpointgc/gc_test.go:
```diff
@@ -0,0 +1,246 @@
+}
+
+func prepareCiliumEndpointCRD(c *k8sClient.FakeClientset) error {
+	c.APIExtFakeClientset.PrependReactor("get", "*", func(action k8sTesting.Action) (handled bool, ret runtime.Object, err error) {
+		return true, nil, nil
+	})
```
Is this needed?

alan-kut replied:
Yes. It passes the:
```go
func (g *GC) checkForCiliumEndpointCRD(ctx hive.HookContext) bool {
	_, err := g.clientset.ApiextensionsV1().CustomResourceDefinitions().Get(
		ctx, k8sconstv2.CEPName, metav1.GetOptions{ResourceVersion: "0"},
	)
	if err == nil {
		return true
	}
        ...
        return false
}
```

pippolo84 reviewed (changes requested):
Thanks for this next iteration! I think there is a fix to be done for an existing operator/api test.

pippolo84 reviewed (approved):
Awesome! Thanks for the contribution! :rocket:
